group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -523,9 +526,84 @@ public CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection
+            CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection> future =
+                runtime.scheduleWriteOperation(
+                    "delete-groups",
+                    topicPartition,
+                    coordinator -> coordinator.deleteGroups(context, groupList)
+                ).exceptionally(exception -> {
+                    if (exception instanceof UnknownTopicOrPartitionException ||
+                        exception instanceof NotEnoughReplicasException) {
+                        return DeleteGroupsRequest.getErrorResultCollection(
+                            groupList,
+                            Errors.COORDINATOR_NOT_AVAILABLE
+                        );
+                    }
+
+                    if (exception instanceof NotLeaderOrFollowerException ||
+                        exception instanceof KafkaStorageException) {
+                        return DeleteGroupsRequest.getErrorResultCollection(
+                            groupList,
+                            Errors.NOT_COORDINATOR
+                        );
+                    }
+
+                    if (exception instanceof RecordTooLargeException ||
+                        exception instanceof RecordBatchTooLargeException ||
+                        exception instanceof InvalidFetchSizeException) {
+                        return DeleteGroupsRequest.getErrorResultCollection(
+                            groupList,
+                            Errors.UNKNOWN_SERVER_ERROR
+                        );
+                    }
+
+                    return DeleteGroupsRequest.getErrorResultCollection(
+                        groupList,
+                        Errors.forException(exception)
+                    );
+                });
+
+            futures.add(future);
+        });
+
+        final CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+        final CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection> resFuture = allFutures.thenApply(v -> {
+            final DeleteGroupsResponseData.DeletableGroupResultCollection res = new DeleteGroupsResponseData.DeletableGroupResultCollection();
+            final List<String> deletedGroups = new ArrayList<>();
+            futures.forEach(future ->
+                // We don't use res.addAll(future.join()) because DeletableGroupResultCollection is an ImplicitLinkedHashMultiCollection,
+                // which has requirements for adding elements (see ImplicitLinkedHashCollection.java#add).
+                future.join().forEach(result -> {
+                    res.add(result.duplicate());
+                    if (result.errorCode() == Errors.NONE.code()) {
+                        deletedGroups.add(result.groupId());
+                    }
+                })
+            );
+            log.info("The following groups were deleted: {}.", String.join(", ", deletedGroups));
```

Comment:
I wonder if we should rather log this within the shard in order to have it logged per shard (with the shard context). What do you think?